runtime-transaction/src/builtin_programs_filter.rs

```diff
@@ -0,0 +1,105 @@
+use {
+    agave_transaction_view::static_account_keys_frame::MAX_STATIC_ACCOUNTS_PER_PACKET as FILTER_SIZE,
+    solana_builtins_default_costs::{is_builtin_program, MAYBE_BUILTIN_KEY},
+    solana_sdk::pubkey::Pubkey,
+};
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub(crate) enum ProgramKind {
+    NotBuiltin,
+    Builtin,
+}
+
+pub(crate) struct BuiltinProgramsFilter {
+    // array of slots for all possible static and sanitized program_id_index,
+    // each slot indicates if a program_id_index has not been checked (eg, None),
+    // or already checked with result (eg, Some(ProgramKind)) that can be reused.
+    program_kind: [Option<ProgramKind>; FILTER_SIZE as usize],
+}
+
+impl BuiltinProgramsFilter {
+    pub(crate) fn new() -> Self {
+        BuiltinProgramsFilter {
+            program_kind: [None; FILTER_SIZE as usize],
+        }
+    }
+
+    pub(crate) fn get_program_kind(&mut self, index: usize, program_id: &Pubkey) -> ProgramKind {
+        *self
+            .program_kind
+            .get_mut(index)
+            .expect("program id index is sanitized")
```

Comment:
do we need expect here? could we return/log an error and handle the error elsewhere by saying this is not a built-in (or vice versa)?

Reply:
`program_kind` is large enough for all static program IDs in a transaction. Any `program_id_index` from a sanitized transaction is expected to be a valid index of `program_kind` array, whose element indicate if it's a builtin or not.